### PR TITLE
feat: github action to check if PR has requested labels before being merged

### PR DESCRIPTION
**Pull request by gauthierpetetin**
**Description**

The purpose of this new Github action is to check if PRs have the requested labels before they get merged:
- A team label, i.e. a label starting with `team-` string (or `external-contributor` label)
- ~~A QA label, i.e. either `QA Passed` or `No QA Needed/E2E Only` or `Spot Check on the Release Build`~~?

Furthermore, it forbids merging PRs with the following labels: `needs-qa`, `QA'd but questions`, `issues-found`, `need-ux-ds-review`, `blocked`, `stale`, `DO-NOT-MERGE`. A consequence is that we can get rid of this [github action](https://github.com/MetaMask/metamask-mobile/blob/main/.github/workflows/ci-all-branches.yml).

**Screenshots/Recordings**

Recording for missing team label: https://recordit.co/U38NvBpIIj
Recording for missing QA label: https://recordit.co/jowmkvMkCh

**Issue**

Progresses https://github.com/MetaMask/mobile-planning/issues/1122

**Checklist**

* [x] There is a related GitHub issue
* [ ] Tests are included if applicable
* [x] Any added code is fully documented

**Other**
- [Same PR for Extension](https://github.com/MetaMask/metamask-extension/pull/19984)

### CODE REVIEW

File: .github/scripts/check-pr-has-required-labels.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  const prNumber = context.payload.pull_request?.number;
+  if (!prNumber) {
+    core.setFailed('Pull request number not found');
+    process.exit(1);
+  }
+
+  // Retrieve pull request labels
+  const prLabels = await retrievePullRequestLabels(octokit, prRepoOwner, prRepoName, prNumber);
+
+  const qaLabels = ["QA Passed", "No QA Needed/E2E Only"];
```
Comment (cortisiko) [inline .github/scripts/check-pr-has-required-labels.ts:33]: I would also add `Spot Check on the Release Build` to this array.

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -6,10 +6,10 @@ Please ensure that any applicable requirements below are satisfied before submit
 
 **Development & PR Process**
 1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
-2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
```
Comment (gauthierpetetin) [inline .github/pull_request_template.md:9]: I removed the requirement for a `release-xx` label as it now added automatically when PR gets merged.

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -6,10 +6,10 @@ Please ensure that any applicable requirements below are satisfied before submit
...
-3. Add `needs-dev-review` label when work is completed
-4. Add `needs-qa` label when dev review is completed
-5. Add `QA Passed` label when QA has signed off
+2. Add `needs-dev-review` label when work is completed
+3. Add `needs-qa` label when dev review is completed
+4. Add `QA Passed` label when QA has signed off
+5. Add your team's label (label starting with `team-`)
```
Comment (gauthierpetetin) [inline .github/pull_request_template.md:12]: I added the requirement for a team label as we now need it to track the number of contributions from other teams

**Review by cortisiko — DISMISSED** [review]
🌮 🌮 🌮

**Comment by gauthierpetetin** [comment]
Hi @cortisiko and @sethkfman ,
I'm not sure how to proceed with this PR as @sethkfman mentioned last Friday, that enforcing the addition of a `QA label` on every PR might introduce a habit where contributors rely on Core Mobile QA team when contributing to Mobile code base, while we want to progressively get away from this.
@cortisiko what do you think?
Does it make sense if I update this PR such that it:
- Forces the addition of a team label, i.e. a label starting with `team-` string (or `external-contributor` label)
- ~~Forces the addition of QA label, i.e. either `QA Passed` or `No QA Needed/E2E Only` or `Spot Check on the Release Build`~~
- Forbids PR to be merged if they have the following labels: `needs-qa`, `QA'd but questions`, `issues-found`.

**Comment by gauthierpetetin** [comment]
@cortisiko @sethkfman, here's a new proposal, where I removed the QA label check in this [commit](https://github.com/MetaMask/metamask-mobile/pull/6794/commits/6864e4b7d68cd9e57a671be87b17e102a2932725).

**Comment by cortisiko** [comment]
@gauthierpetetin Using QA labels on PRs is not limited to the mobile core team; all teams utilize these labels to communicate the type of testing conducted or needed during release testing to the QA engineer. The QA labels help guide us during release testing. We identify based on the label applied to a PR what level of testing is needed during releases. However, if we decide to stop using QA-specific labels, we should have an alternative labeling system to inform the release QA engineers about whether testing has been performed or not. Ideally, engineers would apply this label once testing has been done. For instance, we can use labels to indicate if testing has been completed (automated and or manual), or if the PR doesn't require any testing, such as for documentation or unit test changes.

Anyways what you proposed is fine. We will iron out the details on the proper label soon.

**Comment by gauthierpetetin** [comment]
As discussed with @danjm @sethkfman @cortisiko @PeterYinusa @vpintorico last Friday, I updated this PR to:
- add checks for these labels in this new Github action: `need-ux-ds-review`, `blocked`, `stale`, `DO-NOT-MERGE`
- remove the old [DONOTMERGE Github action]([github action](https://github.com/MetaMask/metamask-mobile/blob/main/.github/workflows/ci-all-branches.yml) as duplicate of this new Github action

Here's the corresponding [commit](https://github.com/MetaMask/metamask-mobile/pull/6794/commits/a1b3fcfc323d773c345794f642a1b7f7aac8f47f).

File: .github/scripts/check-pr-has-required-labels.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      return;
+    }
+  }
+
+  // Otherwise, throw an arror to prevent from merging
+  let errorMessage = '';
+  if (!hasTeamLabel) {
+    errorMessage += 'No team labels found on the PR. ';
+  }
+  errorMessage += 'Please add the required label(s) before merging the PR.';
```
Comment (NicolasMassart) [inline .github/scripts/check-pr-has-required-labels.ts:56]: Suggestion: Maybe add a link in the error message to point to the contribution guide where these labels are explained.

**Review by NicolasMassart — DISMISSED** [review]
Looks good to me. Added only an error message improvement suggestion.

**Review by cortisiko — DISMISSED** [review]
🥇 🥇 🥇

**Comment by Cal-L** [comment]
Updated PR with main

**Review by Cal-L — DISMISSED** [review]
LGTM

File: .github/scripts/check-pr-has-required-labels.ts (hunk 4 above)
Comment (gauthierpetetin) [reply .github/scripts/check-pr-has-required-labels.ts]: Hi @NicolasMassart I didn't want to overload the [CONTRIBUTING.md](https://github.com/MetaMask/metamask-mobile/blob/b7fb3f5e335f19b7d61ea8c491c93649ad36fe06/.github/CONTRIBUTING.md) file, so I created a new file for [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/56d4a712ce5aa3bc43de34e887f9703b57b03143/.github/coding_guidelines/LABELING_GUIDELINES.md) in this [commit](https://github.com/MetaMask/metamask-mobile/pull/6794/commits/56d4a712ce5aa3bc43de34e887f9703b57b03143).

**Review by NicolasMassart — APPROVED** [review]
Looks good to me.

**Pull request merged**